### PR TITLE
add FlashAttention-2 support

### PR DESCRIPTION
**Pull request by iMountTai**
### Description
This PR add support for FlashAttention-2 and enable FlashAttention-2 training by setting the argument `--flash_attn`

**Usage**
Just add the `--flash_attn` to the launching command. For example:
```
torchrun --nnodes 1 --nproc_per_node 1 run_clm_pt_with_peft.py \
    --other_arguments
    --flash_attn
```

### Related Issue
#26

### CODE REVIEW

File: scripts/training/run_clm_pt_with_peft.py [hunk 1]
```diff
@@ -326,6 +327,9 @@ def main():
         model_args, data_args, training_args = parser.parse_json_file(json_file=os.path.abspath(sys.argv[1]))
     else:
         model_args, data_args, training_args = parser.parse_args_into_dataclasses()
+    if training_args.flash_attn:
+        from flash_attn_patch import replace_llama_attn_with_flash_attn
+        replace_llama_attn_with_flash_attn()
```
Comment (airaria) [inline scripts/training/run_clm_pt_with_peft.py:332]: Should it be placed before `import transformers`?

File: scripts/training/run_clm_sft_with_peft.py [hunk 2]
```diff
@@ -213,6 +214,9 @@ def main():
         model_args, data_args, training_args = parser.parse_json_file(json_file=os.path.abspath(sys.argv[1]))
     else:
         model_args, data_args, training_args = parser.parse_args_into_dataclasses()
+    if training_args.flash_attn:
+        from flash_attn_patch import replace_llama_attn_with_flash_attn
+        replace_llama_attn_with_flash_attn()
```
Comment (airaria) [inline scripts/training/run_clm_sft_with_peft.py:219]: Should it be placed before `import transformers`?

File: scripts/training/run_clm_pt_with_peft.py (hunk 1 above)
Comment (airaria) [reply scripts/training/run_clm_pt_with_peft.py:332]: Looks good after discussion.

**Review by airaria — APPROVED** [review]

**Comment by ymcui** [comment]
TODO: Update wiki w.r.t. this PR. @iMountTai

**Pull request merged**